file download does not work on password protected public shares

### Steps to reproduce
1. create a public link share of a folder
2. try to download a file from within that folder via public link

### Expected behaviour
file should be downloaded

### Actual behaviour
file cannot be downloaded


### Client configuration
**Browser:**
Chrome

### Logs
#### Browser log
```
Failed to load resource: the server responded with a status of 401 (Unauthorized)
core.bundle.js:75 Error: Unauthorized
    at core.bundle.js:75```

--- a/apps/files/src/components/FileList.vue
+++ b/apps/files/src/components/FileList.vue
@@ -36,7 +36,15 @@
               </oc-table-cell>
               <oc-table-cell :class="{ 'uk-visible@s' : _sidebarOpen }" class="uk-position-relative">
                 <div class="uk-button-group uk-margin-small-right" :class="{ 'uk-visible@m' : !_sidebarOpen, 'uk-visible@xl' : _sidebarOpen  }">
-                  <oc-button v-for="(action, index) in actions" :key="index" @click.stop="action.handler(item, action.handlerData)" :disabled="!action.isEnabled(item)" :icon="action.icon" :ariaLabel="action.ariaLabel" />
+                  <oc-button
+                    v-for="(action, index) in actions"
+                   :key="index"
+                   @click.stop="action.handler(item, action.handlerData)"
+                   :disabled="!action.isEnabled(item) || $_actionInProgress(item)"
+                   :icon="action.icon"
+                   :ariaLabel="action.ariaLabel"
+                   :uk-tooltip="$_disabledActionTooltip(item)"
+                  />
                 </div>
                 <oc-button
                   :id="'files-file-list-action-button-small-resolution-' + index"
@@ -190,11 +198,24 @@ export default {
 
     enabledActions (item) {
       return this.actions.filter(action => action.isEnabled(item))
+    },
+
+    $_actionInProgress (item) {
+      return this.inProgress.some(itemInProgress => itemInProgress.id === item.id)
+    },
+
+    $_disabledActionTooltip (item) {
+      if (this.$_actionInProgress(item)) {
+        const message = this.$gettext('Another action is currently in progress for this %{itemType}')
+        return this.$gettextInterpolate(message, { itemType: item.type })
+      }
+
+      return null
     }
   },
   computed: {
     ...mapState(['route']),
-    ...mapGetters('Files', ['selectedFiles', 'loadingFolder', 'filesDeleteMessage', 'highlightedFile', 'activeFiles', 'quota', 'filesTotalSize', 'activeFilesCount']),
+    ...mapGetters('Files', ['selectedFiles', 'loadingFolder', 'filesDeleteMessage', 'highlightedFile', 'activeFiles', 'quota', 'filesTotalSize', 'activeFilesCount', 'inProgress']),
     ...mapGetters(['configuration']),
 
     item () {

--- a/apps/files/src/components/FilesAppBar.vue
+++ b/apps/files/src/components/FilesAppBar.vue
@@ -9,6 +9,12 @@
           <span class="uk-text-lead">{{pageTitle}}</span>
         </span>
       </div>
+      <div v-show="inProgress.length > 0" class="uk-width-auto">
+        <oc-spinner id="oc-progress-pie" size="small" />
+        <oc-drop toggle="#oc-progress-pie" mode="click">
+          <upload-menu :items="inProgress" />
+        </oc-drop>
+      </div>
       <div v-if="!publicPage()" class="uk-width-auto uk-visible@m">
         <oc-search-bar @search="onFileSearch" :value="searchTerm" :label="searchLabel" :loading="isLoadingSearch" :button="false"/>
       </div>
@@ -46,12 +52,6 @@
           <oc-button id="oc-filter-list-btn" icon="filter_list" />
           <file-filter-menu />
         </div>
-      </div>
-      <div v-show="inProgress.length > 0" class="uk-width-auto">
-        <oc-spinner id="oc-progress-pie" size="small" />
-        <oc-drop toggle="#oc-progress-pie" mode="click">
-          <upload-menu :items="inProgress" />
-        </oc-drop>
       </div>
     </oc-grid>
     <oc-dialog-prompt name="overwrite-dialog" :oc-active="overwriteDialogMessage !== null" :oc-has-input="false" ocCancelId="files-overwrite-cancel" ocConfirmId="files-overwrite-confirm" :ocTitle="overwriteDialogTitle" :oc-content="overwriteDialogMessage" @oc-confirm="$_ocUpload_confirmOverwrite(true)" @oc-cancel="$_ocUpload_confirmOverwrite(false)" />

--- a/src/plugins/phoenix.js
+++ b/src/plugins/phoenix.js
@@ -1,57 +1,90 @@
-import { mapGetters } from 'vuex'
+import { mapGetters, mapActions } from 'vuex'
 
 export default {
   install (Vue) {
     Vue.mixin({
       computed: {
-        ...mapGetters(['getToken', 'isAuthenticated'])
+        ...mapGetters(['getToken', 'isAuthenticated']),
+        ...mapGetters('Files', ['publicLinkPassword'])
       },
       methods: {
+        ...mapActions('Files', ['addFileToProgress', 'updateFileProgress']),
+        ...mapActions(['showMessage']),
+
         publicPage () {
           return !this.isAuthenticated
         },
         downloadFile (file) {
+          this.addFileToProgress(file)
+          let headers = null
           if (this.publicPage()) {
             const url = this.$client.publicFiles.getFileUrl(file.path)
-            const headers = new Headers()
             const password = this.publicLinkPassword
             if (password) {
-              headers.append('Authorization', 'Basic ' + Buffer.from('public:' + password).toString('base64'))
+              headers = { Authorization: 'Basic ' + Buffer.from('public:' + password).toString('base64') }
             }
 
-            return this.downloadFileFromUrl(url, headers, file.name)
+            return this.downloadFileFromUrl(url, headers, file)
           }
           const url = this.$client.files.getFileUrl(file.path)
-          const headers = new Headers()
-          headers.append('Authorization', 'Bearer ' + this.getToken)
+          headers = { Authorization: 'Bearer ' + this.getToken }
 
-          return this.downloadFileFromUrl(url, headers, file.name)
+          return this.downloadFileFromUrl(url, headers, file)
         },
-        downloadFileFromUrl (url, headers, fileName) {
-          return fetch(url, { headers })
-            .then(response => {
-              if (response.ok) {
-                return response.blob()
-                  .then(blob => {
-                    if (window.navigator && window.navigator.msSaveOrOpenBlob) { // for IE
-                      window.navigator.msSaveOrOpenBlob(blob, fileName)
-                    } else { // for Non-IE (chrome, firefox etc.)
-                      const objectUrl = window.URL.createObjectURL(blob)
+        downloadFileFromUrl (url, headers, file) {
+          const request = new XMLHttpRequest()
+          request.open('GET', url)
+          request.responseType = 'blob'
+          Object.keys(headers).forEach(p => request.setRequestHeader(p, headers[p]))
 
-                      const anchor = document.createElement('a')
-                      anchor.href = objectUrl
-                      anchor.download = fileName
-                      anchor.style.display = 'none'
-                      document.body.appendChild(anchor)
-                      anchor.click()
-                      document.body.removeChild(anchor)
-                      window.URL.revokeObjectURL(objectUrl)
-                    }
-                  })
+          request.addEventListener('readystatechange', () => {
+            if (request.readyState === 2 && request.status !== 200) {
+              this.showMessage({
+                title: this.$gettext('Download failed'),
+                desc: request.statusText,
+                status: 'danger'
+              })
+
+              // Remove item from progress queue in case the request failed
+              this.updateFileProgress({
+                fileName: file.name,
+                progress: 100
+              })
+            } else if (request.readyState === 4 && request.status === 200) {
+              // Download has finished
+              if (window.navigator && window.navigator.msSaveOrOpenBlob) { // for IE
+                window.navigator.msSaveOrOpenBlob(request.response, file.name)
+              } else { // for Non-IE (chrome, firefox etc.)
+                const objectUrl = window.URL.createObjectURL(request.response)
+
+                const anchor = document.createElement('a')
+                anchor.href = objectUrl
+                anchor.download = file.name
+                anchor.style.display = 'none'
+                document.body.appendChild(anchor)
+                anchor.click()
+                document.body.removeChild(anchor)
+                window.URL.revokeObjectURL(objectUrl)
+
+                // Items with small size can be fetched too fast for progress listener
+                // This ensures that they'll be removed from progress queue
+                this.updateFileProgress({
+                  fileName: file.name,
+                  progress: 100
+                })
               }
-              return Promise.reject(new Error(response.statusText))
+            }
+          })
+
+          request.addEventListener('progress', e => {
+            const progress = (e.loaded / e.total) * 100
+            this.updateFileProgress({
+              fileName: file.name,
+              progress
             })
-            .catch(error => console.log(error))
+          })
+
+          request.send()
         }
 
       }

